addons/sale_project/models/sale_order.py

```diff
@@ -377,9 +379,15 @@ def _prepare_invoice_line(self, **optional_values):
                     values['analytic_account_id'] = analytic_account_ids.pop()
         if self.task_id.analytic_tag_ids:
             values['analytic_tag_ids'] += [Command.link(tag_id.id) for tag_id in self.task_id.analytic_tag_ids]
-        elif self.is_service and not self.is_expense:
-            tag_ids = self.env['account.analytic.tag'].search([
-                ('task_ids.sale_line_id', '=', self.id)
-            ])
-            values['analytic_tag_ids'] += [Command.link(tag_id.id) for tag_id in tag_ids]
+        if self.task_id.project_id.analytic_tag_ids:
+            values['analytic_tag_ids'] += [Command.link(tag_id.id) for tag_id in self.task_id.project_id.analytic_tag_ids]
+        if self.is_service and not self.is_expense:
+            domain = []
+            if not self.task_id.analytic_tag_ids:
+                domain = [('task_ids.sale_line_id', '=', self.id)]
+            if not self.task_id.project_id.analytic_tag_ids:
+                domain = expression.OR([[('project_ids.sale_line_id', '=', self.id)], domain])
+            if domain:
+                tag_ids = self.env['account.analytic.tag'].search(domain)
+                values['analytic_tag_ids'] += [Command.link(tag_id.id) for tag_id in tag_ids]
```

Comment:
We could refactor this to have less lines by doing this:
```suggestion
        if self.task_id.analytic_tag_ids or self.task_id.project_id.analytic_tag_ids:
            values['analytic_tag_ids'] += [Command.link(tag_id.id) for tag_id in self.task_id.analytic_tag_ids | self.task_id.project_id.analytic_tag_ids]
        elif self.is_service and not self.is_expense:
            tag_ids = self.env['account.analytic.tag'].search([
                '|', ('task_ids.sale_line_id', '=', self.id), ('project_ids.sale_line_id', '=', self.id)
            ])
            values['analytic_tag_ids'] += [Command.link(tag_id.id) for tag_id in tag_ids]
```

What do you think ? 🤔